stdlib/ballerina-builtin/src/main/java/org/ballerinalang/nativeimpl/reflect/DeepEquals.java

```diff
@@ -0,0 +1,276 @@
+import org.ballerinalang.model.values.BStruct;
+import org.ballerinalang.model.values.BValue;
+import org.ballerinalang.natives.AbstractNativeFunction;
+import org.ballerinalang.natives.annotations.Argument;
+import org.ballerinalang.natives.annotations.BallerinaFunction;
+import org.ballerinalang.natives.annotations.ReturnType;
+
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+/**
+ * The deep equals function which checks equality among 2 values. Currently supports string, int, float, boolean, type,
+ * structs, maps, arrays, any, JSON.
+ */
+@BallerinaFunction(packageName = "ballerina.reflect",
+                   functionName = "deepEquals",
+                   args = {@Argument(name = "value1", type = TypeKind.ANY),
+                           @Argument(name = "value2", type = TypeKind.ANY)},
+                   returnType = {@ReturnType(type = TypeKind.BOOLEAN)},
+                   isPublic = true)
+public class DeepEquals extends AbstractNativeFunction {
+    
+    @Override
+    public BValue[] execute(Context context) {
+        BValue value1 = context.getControlStack().getCurrentFrame().getRefRegs()[0];
+        BValue value2 = context.getControlStack().getCurrentFrame().getRefRegs()[1];
+        boolean isDeepEqual = isDeepEqual(value1, value2);
+        return getBValues(new BBoolean(isDeepEqual));
+    }
+    
+    /**
+     * Deep equal or strict equal implementation for any value type.
+     *
+     * @param lhsValue The value on the left side.
+     * @param rhsValue The value on the right side.
+     * @return True if values are equal, else false.
+     */
+    private boolean isDeepEqual(BValue lhsValue, BValue rhsValue) {
+        if (null == lhsValue && null == rhsValue) {
+            return true;
+        }
+        
+        if (null == lhsValue || null == rhsValue) {
+            return false;
+        }
+        
+        // Required for any == any.
+        if (lhsValue.getType().getTag() != rhsValue.getType().getTag()) {
+            return false;
+        }
+        
+        if (null != lhsValue.getType().getPackagePath() && null != rhsValue.getType().getPackagePath() && !lhsValue
+                .getType().getPackagePath().equals(rhsValue.getType().getPackagePath())) {
+            return false;
+        }
+        
+        if (null != lhsValue.getType().getName() && null != rhsValue.getType().getName() && !lhsValue.getType()
+                .getName().equals(rhsValue.getType().getName())) {
+            return false;
+        }
+        
+        // Handling JSON.
+        if (lhsValue instanceof BJSON && rhsValue instanceof BJSON) {
+            JsonNode lhsJSON = ((BJSON) lhsValue).value();
+            JsonNode rhsJSON = ((BJSON) rhsValue).value();
+            return isDeepEqual(lhsJSON, rhsJSON);
+        }
+        
+        switch (lhsValue.getType().getTag()) {
```

Comment:
This function's body is too big, it should be broken down into more finer logical units. Specially the blocks in the switch/case.

Reply:
Fixed with https://github.com/ballerina-lang/ballerina/pull/4680/commits/0dcdfc6b21f0d6f7bfec05568b490230c5ef6eef